### PR TITLE
Add http proxy support to extension host

### PR DESCRIPTION
**Pull request by code-asher**
For https://github.com/coder/code-server/issues/4672

I tested it with our extension (which downloads the Coder CLI) and confirmed it works.  Once I finish my test extension PR I will add a test for the proxy there as well.

### CODE REVIEW

File: src/vs/platform/request/node/proxy.ts [hunk 1]
```diff
@@ -48,3 +48,22 @@ export async function getProxyAgent(rawRequestURL: string, env: typeof process.e
...
 }
+
+/**
+ * Patch the node http and https modules to route all requests through the agent
+ * we get from the proxy-agent package if a proxy is set in the environment.
+ *
+ * @author coder
+ */
+export async function monkeyPatchHttpAgent(): Promise<void> {
+	if (process.env.http_proxy || process.env.https_proxy || process.env.HTTP_PROXY || process.env.HTTPS_PROXY) {
```
Comment (jsjoeio) [inline src/vs/platform/request/node/proxy.ts:59]: Do we need to add `no_proxy` as an option?

**Review by jsjoeio — APPROVED** [review]
Looks great! No concerns just one clarifying question.

File: src/vs/platform/request/node/proxy.ts (hunk 1 above)
Comment (code-asher) [reply src/vs/platform/request/node/proxy.ts:59]: Nope, `no_proxy` is for specifying exceptions; the proxy agent module will check `no_proxy` when making a request to see whether it should send through the proxy or not.

If `http_proxy` and friends are not set then `no_proxy` has no effect.

**Comment by code-asher** [comment]
Uh...apparently this already works.  I removed my changes to confirm but everything works exactly the same.  I cannot seem to find how they are doing this so I am not sure if I am just doing something dumb like not properly recompiling.

File: src/vs/platform/request/node/proxy.ts (hunk 1 above)
Comment (jsjoeio) [reply src/vs/platform/request/node/proxy.ts:59]: Oh! I see now. Thanks for clarifying!

**Comment by jsjoeio** [comment]
> Uh...apparently this already works. I removed my changes to confirm but everything works exactly the same. I cannot seem to find how they are doing this so I am not sure if I am just doing something dumb like not properly recompiling.

Lol out of curiosity, how are you testing this?

**Comment by code-asher** [comment]
i am running https://github.com/coder/vscode-coder since it makes a download request to fetch the Coder CLI.

Then I run `http_proxy=http://localhost:3128 yarn watch`.  The download fails saying it cannot connect to the proxy.  Then I start the proxy (I am using squid) and it works.

**Comment by code-asher** [comment]
I confirmed it works in the latest artifact built by CI...I still have not located where this happens but I guess for now we are good to move on. 

`no_proxy` does not seem to work but I tested previous versions and trying to use a proxy causes an infinite loop so...it was already broken.

**Comment by jsjoeio** [comment]
Nice thorough testing! Well, I'll leave that issue in code-server open and move it to the next version milestone then!

**Pull request closed without merging**